extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/photos/GooglePhotosExporter.java

```diff
@@ -190,57 +204,39 @@ public GooglePhotosExporter(GoogleCredentialFactory credentialFactory, JobStore
-        .listAlbums(Optional.empty());
-    albumIds.addAll(Arrays.stream(albumListResponse.getAlbums()).map(GoogleAlbum::getId)
-        .collect(Collectors.toList()));
-    String token;
-    while ((token = albumListResponse.getNextPageToken()) != null) {
-      albumListResponse = getOrCreatePhotosInterface(authData)
-          .listAlbums(Optional.of(token));
-      albumIds.addAll(
-          Arrays.stream(albumListResponse
-              .getAlbums())
-              .map(GoogleAlbum::getId)
-              .collect(Collectors.toList()));
-    }
+    // This method is only called once at the beginning of the transfer, so we can start by
+    // initializing a new TempPhotosData to be store in the job store.
+    TempPhotosData tempPhotosData = new TempPhotosData(jobId);
 
-    // Get list of all ids belonging to photos belonging to albums
-    TempPhotosData tempPhotosData = jobStore
-        .findData(jobId, createCacheKey(), TempPhotosData.class);
-    if (tempPhotosData == null) {
-      tempPhotosData = new TempPhotosData(jobId);
-      jobStore.create(jobId, createCacheKey(), tempPhotosData);
-    }
-    for (String albumId : albumIds) {
-      MediaItemSearchResponse albumMediaItemResponse = getOrCreatePhotosInterface(authData)
-          .listMediaItems(Optional.of(albumId), Optional.empty());
-      if (albumMediaItemResponse.getMediaItems().length > 0) {
-        tempPhotosData.addAllContainedPhotoIds(
-            Arrays.stream(albumMediaItemResponse.getMediaItems())
-                .map(GoogleMediaItem::getId)
-                .collect(Collectors.toList()));
-      }
-      while ((token = albumMediaItemResponse.getNextPageToken()) != null) {
-        albumMediaItemResponse = getOrCreatePhotosInterface(authData)
-            .listMediaItems(Optional.of(albumId), Optional.of(token));
-        if (albumMediaItemResponse.getMediaItems().length > 0) {
-          tempPhotosData.addAllContainedPhotoIds(
-              Arrays.stream(albumMediaItemResponse.getMediaItems())
-                  .map(GoogleMediaItem::getId)
-                  .collect(Collectors.toList()));
-        }
+    String albumToken = null;
+    AlbumListResponse albumListResponse;
+    MediaItemSearchResponse containedMediaSearchResponse;
+    do {
+      albumListResponse = getOrCreatePhotosInterface(authData)
+          .listAlbums(Optional.ofNullable(albumToken));
+      for (GoogleAlbum album : albumListResponse.getAlbums()) {
+        String albumId = album.getId();
+        String photoToken = null;
+        do {
+          containedMediaSearchResponse = getOrCreatePhotosInterface(authData)
+              .listMediaItems(Optional.of(albumId), Optional.ofNullable(photoToken));
+          for (GoogleMediaItem mediaItem : containedMediaSearchResponse.getMediaItems()) {
+            tempPhotosData.addContainedPhotoId(mediaItem.getId());
+          }
+        } while ((photoToken = containedMediaSearchResponse.getNextPageToken()) != null);
```

Comment:
I would avoid assignments in conditionals, they get confusing.

Reply:
Done.